File: Common/Csv/CsvParser.vb
Friend Class CsvParser

    Public Sub ParseCsv(ByVal Csv As CsvData, ByVal Str As String, Optional ByVal HasHeaders As Boolean = True, Optional ByVal Delimiter As Char = ","c, Optional ByVal NormalizeLineEndings As Boolean = True)
        Me.Str = Str.ToCharArray()
        Me.Index = 0
        Me.NormalizeLineEndings = True
        Me.DelimiterChar = Delimiter
        Me.Delimiter = Delimiter
        Me.Csv = Csv

        Me.ParseCsv(HasHeaders)
    End Sub

    Private Shared Sub EnsureColumns(ByVal Data As CsvData, ByVal I As Integer)
        Do Until I < Data.Columns.Count
            Data.Columns.Add()
        Loop
    End Sub

    Private Sub ParseCsv(ByVal HasHeaders As Boolean)
        Dim Res = Me.Csv
        Res.Clear()

        Res.HasHeaders = HasHeaders
        If HasHeaders Then
            Do
                Dim T = Me.ReadToken()
                Dim D = T
                Dim ShouldPeek = True

                ' If a delimiter happens right after another, then an empty field should be considered between them.
                If T Is NewLine Or T Is Nothing Or T Is Delimiter Then
                    D = ""
                    ShouldPeek = False
                End If

                Res.Columns.Add(D)

                If ShouldPeek Then
                    T = Me.PeekToken()
                End If

                If T Is NewLine Or T Is Nothing Or T Is Delimiter Then
                    If ShouldPeek Then
                        Me.ReadToken()
                    End If
                    If T Is Delimiter Then
                        Continue Do
                    Else
                        Exit Do
                    End If
                End If

                Throw New ArgumentException("Invalid CSV.")
            Loop
        End If

        Dim Entry As CsvEntry = Nothing
        Dim IsLastEntryEmpty As Boolean
        Dim I As Integer
        Dim NewEntry = True

        Do
            If NewEntry Then
                Entry = Res.Entries.Add()
                I = 0
                IsLastEntryEmpty = True
                NewEntry = False
            End If

            Dim T = Me.ReadToken()
            Dim D = T
            Dim ShouldPeek = True

            ' The last empty line. We have to remove the last entry from the data.
            If T Is Nothing And IsLastEntryEmpty Then
                Exit Do
            End If

            ' If a delimiter happens right after another, then an empty field should be considered between them.
            If T Is Nothing Or T Is NewLine Or T Is Delimiter Then
                D = ""
                ShouldPeek = False
            End If

            EnsureColumns(Res, I)
            Entry.Item(I) = D
            I += 1
            IsLastEntryEmpty = False

            If ShouldPeek Then
                T = Me.PeekToken()
            End If

            If T Is Nothing Then
                Exit Do
            End If
            If T Is NewLine Or T Is Delimiter Then
                If ShouldPeek Then
                    Me.ReadToken()
                End If
                If T Is NewLine Then
                    NewEntry = True
                End If
                Continue Do
            End If

            Throw New ArgumentException("Invalid CSV.")
        Loop

        If IsLastEntryEmpty Then
            Res.Entries.Remove(Res.Entries.Count - 1)
        End If
    End Sub

    Private Function ReadToken(ByVal T As String) As Boolean
        If Me.PeekToken() = T Then
            Me.ReadToken()
            Return True
        End If
        Return False
    End Function

    Private Function PeekToken() As String
        Dim I = Me.Index
        Dim R = Me.ReadToken()
        Me.Index = I
        Return R
    End Function

    ''' <summary>
    ''' Possible Values:
    ''' Nothing
    ''' NewLine
    ''' ","
    ''' Field
    ''' </summary>
    Private Function ReadToken() As String
        If Me.IsFinished() Then
            Return Nothing
        End If

        If Me.ReadChar(ControlChars.Cr) Then
            Me.ReadChar(ControlChars.Lf)
            Return NewLine
        End If

        If Me.ReadChar(ControlChars.Lf) Then
            Return NewLine
        End If

        If Me.ReadChar(Me.DelimiterChar) Then
            Return Me.Delimiter
        End If

        Dim Res = New Text.StringBuilder()

        If Me.ReadChar(""""c) Then
            Do
                If Me.IsFinished() Then
                    Throw New ArgumentException("Invalid CSV.")
                End If

                If Me.ReadChar(""""c) Then
                    If Not Me.ReadChar(""""c) Then
                        Exit Do
                    End If
                    Res.Append(""""c)
                    Continue Do
                End If

                If Me.NormalizeLineEndings AndAlso Me.ReadChar(ControlChars.Cr) Then
                    Me.ReadChar(ControlChars.Lf)
                    Res.Append(NewLine)
                    Continue Do
                End If
                If Me.NormalizeLineEndings AndAlso Me.ReadChar(ControlChars.Lf) Then
                    Res.Append(NewLine)
                    Continue Do
                End If

                Res.Append(Me.ReadChar())
            Loop
        Else
            Do
                Dim T = Me.PeekChar()
                If Not T.HasValue Or T = Me.DelimiterChar Or T = ControlChars.Cr Or T = ControlChars.Lf Then
                    Exit Do
                End If
                Res.Append(Me.ReadChar())
            Loop
        End If

        Return Res.ToString()
    End Function

    Private Function ReadChar(ByVal Ch As Char) As Boolean
        If Me.PeekChar() = Ch Then
            Me.ReadChar()
            Return True
        End If
        Return False
    End Function

    Private Function ReadChar() As Char?
        If Me.IsFinished() Then
            Return Nothing
        End If
        Dim R = Me.Str(Index)
        Index += 1
        Return R
    End Function

    Private Function PeekChar() As Char?
        If Me.IsFinished() Then
            Return Nothing
        End If
        Return Me.Str(Index)
    End Function

    Private Function IsFinished() As Boolean
        Return Me.Index = Me.Str.Length
    End Function

#Region "Instance Property"
    Private Shared ReadOnly _Instance As CsvParser = New CsvParser()

    Public Shared ReadOnly Property Instance As CsvParser
        Get
            Return _Instance
        End Get
    End Property
#End Region

    Private Shared NewLine As String = ControlChars.CrLf

    Private Delimiter As String
    Private DelimiterChar As Char
    Private NormalizeLineEndings As Boolean
    Private Str As Char()
    Private Index As Integer
    Private Csv As CsvData

End Class

File: Common/Csv/CsvParser.cs
using System;

namespace Ks.Common
{
    internal class CsvParser
    {
        public void ParseCsv(CsvData Csv, string Str, bool HasHeaders = true, char Delimiter = ',', bool NormalizeLineEndings = true)
        {
            this.Str = Str.ToCharArray();
            this.Index = 0;
            this.NormalizeLineEndings = true;
            this.DelimiterChar = Delimiter;
            this.Delimiter = Delimiter.ToString();
            this.Csv = Csv;

            this.ParseCsv(HasHeaders);
        }

        private static void EnsureColumns(CsvData Data, int I)
        {
            while (I >= Data.Columns.Count)
                Data.Columns.Add();
        }

        private void ParseCsv(bool HasHeaders)
        {
            var Res = this.Csv;
            Res.Clear();

            Res.HasHeaders = HasHeaders;
            if (HasHeaders)
            {
                while (true)
                {
                    var T = this.ReadToken();
                    var D = T;
                    var ShouldPeek = true;

                    // If a delimiter happens right after another, then an empty field should be considered between them.
                    if ((T == NewLine) | T == null | (T == Delimiter))
                    {
                        D = "";
                        ShouldPeek = false;
                    }

                    Res.Columns.Add(D);
                    if (ShouldPeek)
                        T = this.PeekToken();
                    if ((T == NewLine) | T == null | (T == Delimiter))
                    {
                        if (ShouldPeek)
                            this.ReadToken();
                        if (T == Delimiter)
                            continue;
                        else
                            break;
                    }

                    throw new ArgumentException("Invalid CSV.");
                }
            }

            CsvEntry Entry = null;
            var IsLastEntryEmpty = false;
            var I = 0;
            var NewEntry = true;

            while (true)
            {
                if (NewEntry)
                {
                    Entry = Res.Entries.Add();
                    I = 0;
                    IsLastEntryEmpty = true;
                    NewEntry = false;
                }

                var T = this.ReadToken();
                var D = T;
                var ShouldPeek = true;

                // The last empty line. We have to remove the last entry from the data.
                if (T == null & IsLastEntryEmpty)
                    break;

                // If a delimiter happens right after another, then an empty field should be considered between them.
                if (T == null | (T == NewLine) | (T == Delimiter))
                {
                    D = "";
                    ShouldPeek = false;
                }

                EnsureColumns(Res, I);
                Entry[I] = D;
                I += 1;
                IsLastEntryEmpty = false;
                if (ShouldPeek)
                    T = this.PeekToken();
                if (T == null)
                    break;
                if ((T == NewLine) | (T == Delimiter))
                {
                    if (ShouldPeek)
                        this.ReadToken();
                    if (T == NewLine)
                        NewEntry = true;
                    continue;
                }

                throw new ArgumentException("Invalid CSV.");
            }

            if (IsLastEntryEmpty)
                Res.Entries.Remove(Res.Entries.Count - 1);
        }

        private bool ReadToken(string T)
        {
            if (this.PeekToken() == T)
            {
                this.ReadToken();
                return true;
            }
            return false;
        }

        private string PeekToken()
        {
            var I = this.Index;
            var R = this.ReadToken();
            this.Index = I;
            return R;
        }

        /// <summary>
        /// Possible Values:
        /// Nothing
        /// NewLine
        /// ","
        /// Field
        /// </summary>
        private string ReadToken()
        {
            if (this.IsFinished())
                return null;

            if (this.ReadChar('\r'))
            {
                this.ReadChar('\n');
                return NewLine;
            }

            if (this.ReadChar('\n'))
                return NewLine;

            if (this.ReadChar(this.DelimiterChar))
                return this.Delimiter;

            var Res = new System.Text.StringBuilder();

            if (this.ReadChar('"'))
            {
                while (true)
                {
                    if (this.IsFinished())
                        throw new ArgumentException("Invalid CSV.");
                    if (this.ReadChar('"'))
                    {
                        if (!this.ReadChar('"'))
                            break;
                        Res.Append('"');
                        continue;
                    }

                    if (this.NormalizeLineEndings && this.ReadChar('\r'))
                    {
                        this.ReadChar('\n');
                        Res.Append(NewLine);
                        continue;
                    }

                    if (this.NormalizeLineEndings && this.ReadChar('\n'))
                    {
                        Res.Append(NewLine);
                        continue;
                    }

                    Res.Append(this.ReadChar());
                }
            }
            else
                while (true)
                {
                    var T = this.PeekChar();
                    if (!T.HasValue | (T == this.DelimiterChar) | (T == '\r') | (T == '\n'))
                        break;
                    Res.Append(this.ReadChar());
                }

            return Res.ToString();
        }

        private bool ReadChar(char Ch)
        {
            if (this.PeekChar() == Ch)
            {
                this.ReadChar();
                return true;
            }
            return false;
        }

        private char? ReadChar()
        {
            if (this.IsFinished())
                return default;
            var R = this.Str[Index];
            Index += 1;
            return R;
        }

        private char? PeekChar()
        {
            if (this.IsFinished())
                return default;
            return this.Str[Index];
        }

        private bool IsFinished()
        {
            return this.Index == this.Str.Length;
        }

        public static CsvParser Instance { get; } = new CsvParser();

        private static string NewLine = "\r\n";

        private string Delimiter;
        private char DelimiterChar;
        private bool NormalizeLineEndings;
        private char[] Str;
        private int Index;
        private CsvData Csv;
    }
}
